slf4j/src/main/java/com/google/common/flogger/backend/slf4j/Slf4jLoggerBackend.java

```diff
@@ -0,0 +1,179 @@
+import com.google.common.flogger.backend.SimpleMessageFormatter;
+import org.slf4j.Logger;
+
+import java.util.logging.Level;
+
+/**
+ * A logging backend that uses slf4j to output log statements.
+ */
+final class Slf4jLoggerBackend extends LoggerBackend implements
+    SimpleMessageFormatter.SimpleLogHandler {
+
+  private final Logger logger;
+
+  private static final MappedLevel TRACE_LEVEL = new MappedLevel(new MappedLevel.Predicate() {
+
+    @Override
+    public boolean test(Logger logger) {
+      return logger.isTraceEnabled();
+    }
+  }, new MappedLevel.LogAdapter() {
+    @Override
+    public void log(Logger logger, String message, Throwable thrown) {
+      logger.trace(message, thrown);
+    }
+  });
+
+  private static final MappedLevel DEBUG_LEVEL = new MappedLevel(new MappedLevel.Predicate() {
+
+    @Override
+    public boolean test(Logger logger) {
+      return logger.isDebugEnabled();
+    }
+  }, new MappedLevel.LogAdapter() {
+    @Override
+    public void log(Logger logger, String message, Throwable thrown) {
+      logger.debug(message, thrown);
+    }
+  });
+  private static final MappedLevel INFO_LEVEL = new MappedLevel(new MappedLevel.Predicate() {
+
+    @Override
+    public boolean test(Logger logger) {
+      return logger.isInfoEnabled();
+    }
+  }, new MappedLevel.LogAdapter() {
+    @Override
+    public void log(Logger logger, String message, Throwable thrown) {
+      logger.info(message, thrown);
+    }
+  });
+  private static final MappedLevel WARN_LEVEL = new MappedLevel(new MappedLevel.Predicate() {
+
+    @Override
+    public boolean test(Logger logger) {
+      return logger.isWarnEnabled();
+    }
+  }, new MappedLevel.LogAdapter() {
+    @Override
+    public void log(Logger logger, String message, Throwable thrown) {
+      logger.warn(message, thrown);
+    }
+  });
+  private static final MappedLevel ERROR_LEVEL = new MappedLevel(new MappedLevel.Predicate() {
+
+    @Override
+    public boolean test(Logger logger) {
+      return logger.isErrorEnabled();
+    }
+  }, new MappedLevel.LogAdapter() {
+    @Override
+    public void log(Logger logger, String message, Throwable thrown) {
+      logger.error(message, thrown);
+    }
+  });
+
+  private static final MappedLevel[] mappedLevels = new MappedLevel[Level.SEVERE.intValue() + 1];
+
+  static {
+    mappedLevels[Level.FINEST.intValue()] = TRACE_LEVEL;
+    mappedLevels[Level.FINER.intValue()] = TRACE_LEVEL;
+    mappedLevels[Level.FINE.intValue()] = DEBUG_LEVEL;
+    mappedLevels[Level.CONFIG.intValue()] = DEBUG_LEVEL;
+    mappedLevels[Level.INFO.intValue()] = INFO_LEVEL;
+    mappedLevels[Level.WARNING.intValue()] = WARN_LEVEL;
+    mappedLevels[Level.SEVERE.intValue()] = ERROR_LEVEL;
+  }
+
+  Slf4jLoggerBackend(Logger logger) {
+    if (logger == null) {
+      throw new NullPointerException("logger is required");
+    }
+    this.logger = logger;
+  }
+
+  private static MappedLevel mapLevel(Level level) {
```

Comment:
Does this work if someone calls logger.at(MY_CUSTOM_LOG_LEVEL).log("whatever"); ??
It's not safe to assume that given levels are only those defined in the JDK.

Reply:
Pushed commit allowing for custom levels to be safely handled.

Reply:
I have to say I'm still not sure I understand the need for the handler classes. Making closures like that which capture some functionality works well if you get in and cache it, since that hides the cost of the lookup on every event.

However you are resolving the MappedLevel instance every time and then using it immediately without storing it.
Why not just put the logic inside the two places you need it. You could delete about 100 lines from this file (including the MappedLevel class) and replace it with 2 similar looking if-chains or switch statements (if you went via an enum for the SLF4J level, which might be worth it for readability).

The MappedLevel class just doesn't seem like a useful abstraction and it's not really saving any work (while it does make the code a bit harder to grok in my opinion).

Reply:
Agreed; it was intended to avoid duplicating the mapping logic.  Pushed a commit that adds SLF4J log level enumeration, separating the JUL -> SLF4J mapping from dispatch (which is now two simple switch statements).